firecracker-control/local.go

```diff
@@ -0,0 +1,319 @@
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type: plugin.ServicePlugin,
+		ID:   localPluginID,
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			log.G(ic.Context).Debugf("initializing %s plugin (root: %q)", localPluginID, ic.Root)
+			return newLocal(ic)
+		},
+	})
+}
+
+// TODO: Remove this interface after merging PR #91
+// See https://github.com/firecracker-microvm/firecracker-go-sdk/pull/91
+type machineIface interface {
+	Start(context.Context) error
+	StopVMM() error
+	Wait(context.Context) error
+	SetMetadata(context.Context, interface{}) error
+}
+
+type instance struct {
+	cfg     *firecracker.Config
+	machine machineIface
+}
+
+type local struct {
+	vm          map[string]*instance
+	vmLock      sync.RWMutex
+	rootPath    string
+	findVsockFn func(context.Context) (uint32, error)
+}
+
+func newLocal(ic *plugin.InitContext) (*local, error) {
+	if err := os.MkdirAll(ic.Root, 0750); err != nil && !os.IsExist(err) {
+		return nil, errors.Wrapf(err, "failed to create root directory: %s", ic.Root)
+	}
+
+	return &local{
+		vm:          make(map[string]*instance),
+		rootPath:    ic.Root,
+		findVsockFn: findNextAvailableVsockCID,
+	}, nil
+}
+
+// CreateVM creates new Firecracker VM instance
+func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*proto.CreateVMResponse, error) {
+	id, err := s.makeID()
+	if err != nil {
+		return nil, err
+	}
+
+	logger := log.G(ctx).WithField("vm_id", id)
+
+	cfg, err := s.buildVMConfiguration(ctx, id, req)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("creating new VM")
+
+	machine, err := firecracker.NewMachine(ctx, *cfg, firecracker.WithLogger(logger))
+	if err != nil {
+		logger.WithError(err).Error("failed to create new machine instance")
+		return nil, err
+	}
+
+	if err := s.startMachine(ctx, id, machine); err != nil {
+		logger.WithError(err).Error("failed to start VM instance")
+		return nil, err
+	}
+
+	logger.Info("successfully started the VM")
+
+	newInstance := &instance{
+		cfg:     cfg,
+		machine: machine,
+	}
+
+	s.vmLock.Lock()
+	s.vm[id] = newInstance
+	s.vmLock.Unlock()
+
+	return &proto.CreateVMResponse{
+		VMID: id,
+	}, nil
+}
+
+func (s *local) makeID() (string, error) {
+	number, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt32))
+	if err != nil {
+		return "", err
+	}
+
+	return strconv.FormatInt(number.Int64(), 10), nil
+}
+
+func (s *local) buildVMConfiguration(ctx context.Context, id string, req *proto.CreateVMRequest) (*firecracker.Config, error) {
+	logger := log.G(ctx).WithField("vm_id", id)
+
+	cid, err := s.findVsockFn(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find available cid for vsock")
+	}
+
+	vsockDevices := []firecracker.VsockDevice{
+		{Path: "root", CID: cid},
+	}
+
+	logger.Debugf("using cid: %d", cid)
+
+	var (
+		socketPath      = filepath.Join(s.rootPath, fmt.Sprintf("vm_%s.socket", id))
+		logFifoPath     = filepath.Join(s.rootPath, fmt.Sprintf("vm_%s_log.fifo", id))
```

Comment:
Is `s.rootPath` different for every VM? If not, won't all VMs be using conflicting file paths for the socket and FIFOs?

Reply:
`rootPath` is assigned by containerd as a root directory for a plugin.
There will be no conflicts as long as ids are unique for each VM.

Reply:
Ah, I overlooked the fact that each file itself is parameterized by VMID, for some reason I was just imagining we'd make one directory per-VM. Makes sense